feat: Added check for existing package and a new --overwrite-dest CLI option in build command

harikishen opened the pull request:
Fixes https://github.com/mozilla/web-ext/issues/68

Done some changes to fix the issue

rpl commented on src/cmd/build.js:
```diff
@@ -124,11 +130,42 @@ async function defaultPackageCreator({
+    destStat = await fs.stat(extensionPath);
+  } catch (error) {
+    destStat = false;
+    throw error;
```
@harikishen it looks like the existent tests are currently failing because fs.stat raises an error when the "extensionPath" does not exist (it raises an error object with code 'ENOENT', e.g. like documented in the following nodejs doc pages: https://nodejs.org/api/errors.html#errors_common_system_errors and https://nodejs.org/api/fs.html#fs_fs_stat_path_callback) and so we should throw the caught error only if the error code is not 'ENOENT' (which is an expected error when we are building a file to a path that doesn't exist yet).

Applying the following change should fix the existent tests:
```diff
diff --git a/src/cmd/build.js b/src/cmd/build.js
index 125eed7..6d123d2 100644
--- a/src/cmd/build.js
+++ b/src/cmd/build.js
@@ -11,7 +11,7 @@ import {zipDir} from '../util/zip-dir';
 import getValidatedManifest, {getManifestId} from '../util/manifest';
 import {prepareArtifactsDir} from '../util/artifacts';
 import {createLogger} from '../util/logger';
-import {UsageError} from '../errors';
+import {UsageError, isErrorWithCode} from '../errors';
 import {
   createFileFilter as defaultFileFilterCreator,
   FileFilter,
@@ -135,7 +135,9 @@ async function defaultPackageCreator({
     destStat = await fs.stat(extensionPath);
   } catch (error) {
     destStat = false;
-    throw error;
+    if (!isErrorWithCode('ENOENT', error)) {
+      throw error;
+    }
   }
   const destinationFileExists = destStat && destStat.isFile();
```

rpl commented on src/cmd/build.js:
```diff
@@ -124,11 +130,42 @@ async function defaultPackageCreator({
+  }
+
+  if (destinationFileExists) {
+    if (overwriteDest) {
```
Instead of nesting two `if` statement into each other we can probably do something like:

```js
if (!overwriteDest && destinationFileExists) {
  throw new UsageError(...);
}

if (destinationFileExists) {
  log.info(`Destination exists, ...`);
}

...
```

harikishen replied:
but won't both conditions be satisfied if this is true `if (!overwriteDest && destinationFileExists)`

harikishen commented:
@rpl @kumar303 I've added the `--overwrite-dest` option and the check for existing packages in [build.js](https://github.com/harikishen/web-ext/blob/ticket_68/src/cmd/build.js) and [program.js](https://github.com/harikishen/web-ext/blob/ticket_68/src/program.js). Also i have added the corresponding [unit tests](https://github.com/harikishen/web-ext/blob/ticket_68/tests/unit/test-cmd/test.build.js)

rpl commented:
@harikishen hehe no worries, the build on travis is failing because of the linting checks on the pull request title (which becomes the merged commit message if the PR is composed by more then one commit, and given that we generate the changelog using conventional-changelog, we run [conventinal-changelog-lint](https://github.com/marionebl/conventional-changelog-lint) on every pull request update to prevent us to merge it with a commit message that would make the changelog generation to fail once we release a new web-ext version).

I just fixed the PR title and restarted the travis job, which is now green.

Coveralls is still red, though, which means that there is still something that needs to be tested explicitly.

rpl commented:
I just looked into the Coveralls report and it looks that the following are the scenarios that are missing a test:

- `await fs.stat(extensionPath)` raises an error that is not an error with code `"ENOENT"` (https://coveralls.io/builds/11654589/source?filename=src%2Fcmd%2Fbuild.js#L138)
- `await eventToPromise(stream, 'close')` raises an error with code `"EEXIST"` (https://coveralls.io/builds/11654589/source?filename=src%2Fcmd%2Fbuild.js#L167)

While the first one can probably be reproduced by forcing build to use a directory where we don't have write permissions (e.g. we used this strategy in this test file: https://github.com/mozilla/web-ext/blob/master/tests/unit/test-util/test.artifacts.js#L60-L123), but in that case we would need to skip the test on windows, on the contrary the build will fail on appveyor (the CI service that we use to run the tests on windows)

The second one is harder, because it would require the test to recreate the race condition of something that create the file between the fs.stat and the actual file creation.

Actually, I'm not sure that we should check that the file exist after we did the `fs.stat` call, because if I get the feature correctly, it should prevent the developer to overwrite a file that is already there, and not protecting the user from the issue that could arise by running the build concurrently on more than one shell, am I right?

@kumar303 what to you think about it?

harikishen commented:
@rpl i think this line maybe the cause for the issue https://github.com/harikishen/web-ext/blob/ticket_68/src/cmd/build.js#L167 because i didn't follow the steps given by @kumar303 in the reference pull request

rpl commented:
@harikishen another reasonable way to approach it could be the following:

- do not check the file existence with `fs.stat`, but just use `createWriteStream(extensionPath, {flags: 'wx'})` the first time, so that it raises an exception if the file already exist (the error with code 'EEXIST')
- if the file already exist and the `overwriteDest` option has been specified, rerun `createWriteStream` without the 'wx' flags, or raise the exception if the `overwriteDest` option has not been enabled.

Something like the following:
```
  const stream = createWriteStream(extensionPath, {flags: 'wx'}); // Check if destination file exists first

  stream.write(buffer, () => stream.end());

  try {
    await eventToPromise(stream, 'close');
  } catch (error) {
    if (error.code === 'EEXIST') {
      if (overwriteDest) {
        log.info(`Destination exists, overwriting: ${extensionPath}`);

        // Retry, this time by overwriting the existent file.
        const overwriteStream = createWriteStream(extensionPath);
        stream.write(bugger, () => stream.end());
        await eventToPromise(stream, 'close');
      } else {
        throw new UsageError(
          `Extension exists at the destination path: ${extensionPath}\n` +
            'Use --overwrite-dest to enable overwriting.');
      }
    } else {
       // Re-throw any unexpected exception as it is.
       throw error;
    }
  }
```

This way we achieve the same safety of the current implementation, but with a simplified implementation.

How that sounds to you?

harikishen commented:
actually i thought of that the first time i saw the code. but @kumar303 mentioned on performing the `fs.stat` check somewhere in the previous PR

rpl commented:
@harikishen true @kumar303 mentioned `fs.stat` in the issue comments, but while I've been looking at the code in this PR again this morning and thinking about the scenarios, I realized that the first `fs.stat` call doesn't make the "file existence check" safer, it just make it unnecessarily more complex (and create more code branches to cover with the unit tests).

Given that using the 'wx' flags makes the "file existence check" pretty safe, we can use it as the only "file existence check", which will also simplify its testing.

kumar303 commented:
> This way we achieve the same safety of the current implementation, but with a simplified implementation.

I agree with the approach illustrated above. Making less code paths is always better.

harikishen commented:
@kumar303  @rpl  I've made the said changes.

rpl commented on src/cmd/build.js:
```diff
@@ -124,11 +130,31 @@ async function defaultPackageCreator({
+  } catch (error) {
+    if (isErrorWithCode('EEXIST', error)) {
+      throwExtensionExists();
+    }
```
we should re-raise any unexpected error here (basically anything that is not an 'EEXIST' error.

rpl commented on src/cmd/build.js:
```diff
@@ -4,14 +4,14 @@ import {createWriteStream} from 'fs';
 import {fs} from 'mz';
 import parseJSON from 'parse-json';
-import eventToPromise from 'event-to-promise';
+import {eventToPromise as defaultEventToPromise} from 'event-to-promise';
```
```js
import defaultEventToPromise from 'event-to-promise';
```

rpl replied:
@harikishen you forgot to apply this change in the last update on this patch.

`eventToPromise` is not a named export of the 'event-to-promise' npm package, it is its default export and it should be imported with `defaultEventToPromise` as its name using the syntax:
```js
import defaultEventToPromise from 'event-to-promise';
```

rpl commented on src/cmd/build.js:
```diff
@@ -40,6 +40,8 @@ export type PackageCreatorParams = {|
   sourceDir: string,
   fileFilter: FileFilter,
   artifactsDir: string,
+  eventToPromise: typeof defaultEventToPromise,
```
This is an injected dependency and by convention we are always separating the regular parameter (with the options that the function is supposed to get usually) from the injected parameters (which are usually redefined only by the unit tests).

And so, `eventToPromise` should be removed from the `PackageCreatorParams` and a new `PackageCreatorOptions` type should be created:

```js
export type PackageCreatorOptions = {|
  eventToPromise: typeof defaultEventToPromise,
|};
```

rpl commented on src/cmd/build.js:
```diff
@@ -97,8 +99,14 @@ export async function getDefaultLocalizedName(
 
-async function defaultPackageCreator({
-  manifestData, sourceDir, fileFilter, artifactsDir, showReadyMessage,
+export async function defaultPackageCreator({
```
Like mentioned in the previous comment, eventToPromise should be in a second parameter, not in the first one.

```js
export async function defaultPackageCreator(
  {
    manifestData,
    sourceDir,
    fileFilter,
    artifactsDir,
    overwriteDest,
    showReadyMessage,
  }: PackageCreatorParams,
  {
    eventToPromise = defaultEventToPromise,
  }: PackageCreatorOptions = {}
): Promise<ExtensionBuildResult> {
  ... /* here is where the function body is */
}
```

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +368,27 @@ describe('build', () => {
+        sourceDir: fixturePath('minimal-web-ext'),
+        fileFilter: {wantFile: () => true},
+        artifactsDir: 'artifacts',
+        eventToPromise: fakeEventToPromise,
```
Once you have applied the changes to the "src/cmd/build.js" module described above, we can fix this test accordingly (by moving the fake `eventToPromise` in the second parameter, and by using a real FileFilter instance to prevent flow to complain at line 385):

```diff
diff --git a/tests/unit/test-cmd/test.build.js b/tests/unit/test-cmd/test.build.js
index 14ce59d..ce54181 100644
--- a/tests/unit/test-cmd/test.build.js
+++ b/tests/unit/test-cmd/test.build.js
@@ -373,16 +373,23 @@ describe('build', () => {
     it('should reject on Unexpected errors', () => {
       const fakeEventToPromise = sinon.spy(
         () => Promise.reject(new Error('Unexpected error')));
+      const sourceDir = fixturePath('minimal-web-ext');
+      const artifactsDir = 'artifacts';
+      const fileFilter = new FileFilter({sourceDir, artifactsDir});
       const params = {
         manifestData: basicManifest,
-        sourceDir: fixturePath('minimal-web-ext'),
-        fileFilter: {wantFile: () => true},
-        artifactsDir: 'artifacts',
-        eventToPromise: fakeEventToPromise,
+        sourceDir,
+        fileFilter,
+        artifactsDir,
         overwriteDest: false,
         showReadyMessage: false,
-      };// $FLOW_FIXME: flow complains about the fileFiler type
-      return defaultPackageCreator(params)
+      };
+
+      const options = {
+        eventToPromise: fakeEventToPromise,
+      };
+
+      return defaultPackageCreator(params, options)
         .then(() => assert.fail('The expected error has not been raised'))
         .catch ((error) => {
           assert.match(error.message, /Unexpected error/);
```

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +321,44 @@ describe('build', () => {
+      (tmpDir) => {
+        const testFileName =
+          path.join(tmpDir.path(), 'minimal_extension-1.0.zip');
+        return fs.writeFile(testFileName, 'test')
```
@harikishen We need to remove this file to fix this issue when this test run on Windows:
https://ci.appveyor.com/project/kumar303/web-ext/build/1.0.518#L1693

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +321,44 @@ describe('build', () => {
+          path.join(tmpDir.path(), 'minimal_extension-1.0.zip');
+        return fs.writeFile(testFileName, 'test')
+          .then(() =>
+        build({
```
@harikishen can you indent this `build({...})` a bit to make this lines nicer to read?

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -376,8 +373,12 @@ describe('build', () => {
+        const fakeEventToPromise = sinon.spy(async (stream) => {
+          await defaultEventToPromise(stream, 'close');
+          await fs.unlink(
+            path.join(tmpDir.path(), 'minimal_extension-1.0.zip'));
```
This test fails because the created zip file is not named 'minimal_extension-1.0.zip'.

Let's delete the file by discovering its name inside this async function, so that if the file name changes
the test is still able to remove the tmpDir path correctly:

```diff
diff --git a/tests/unit/test-cmd/test.build.js b/tests/unit/test-cmd/test.build.js
index ca6e0e8..b483057 100644
--- a/tests/unit/test-cmd/test.build.js
+++ b/tests/unit/test-cmd/test.build.js
@@ -375,8 +375,10 @@ describe('build', () => {
       (tmpDir) => {
         const fakeEventToPromise = sinon.spy(async (stream) => {
           await defaultEventToPromise(stream, 'close');
-          await fs.unlink(
-            path.join(tmpDir.path(), 'minimal_extension-1.0.zip'));
+          const files = await fs.readdir(tmpDir.path());
+          for (const file of files) {
+            await fs.unlink(path.join(tmpDir.path(), file));
+          }
           return Promise.reject(new Error('Unexpected error'));
         });
         const sourceDir = fixturePath('minimal-web-ext');
```

rpl commented on src/cmd/build.js:
```diff
@@ -97,9 +102,19 @@ export async function getDefaultLocalizedName(
+  }: PackageCreatorParams,
+  {
+    eventToPromise = defaultEventToPromise,
+  }:PackageCreatorOptions = {}
```
This line has a lint warning (a space is missing between `:` and `PackageCreatorOptions`):

```
$ npm run lint
> grunt lint

Running "lint" task

Running "eslint:js" (eslint) task

/.../web-ext/src/cmd/build.js
  114:3  warning  There must be a space after "{
    eventToPromise = defaultEventToPromise,
  }:PackageCreatorOptions" parameter type annotation colon  flowtype/space-after-type-colon

✖ 1 problem (0 errors, 1 warning)
```

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,44 @@ describe('build', () => {
+          path.join(tmpDir.path(), 'minimal_extension-1.0.zip');
+        return fs.writeFile(testFileName, 'test')
+          .then(() =>
+            build({
```
can you move `build({` to the previous line? it makes much more clear that the arrow function is going to return the result of the `build` call.

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,44 @@ describe('build', () => {
+          .catch ((error) => {
+            assert.instanceOf(error, UsageError);
+          });
+
```
Nit: remove this empty line

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,44 @@ describe('build', () => {
+  it('overwrites zip file if it exists', () => {
+    return withTempDir(
+      (tmpDir) => {
+        const testFileName =
```
same coding style issue mentioned in the previous comment.

rpl replied:
@harikishen this one is not yet fixed.

harikishen replied:
@rpl i've changed it

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,44 @@ describe('build', () => {
+          path.join(tmpDir.path(), 'minimal_extension-1.0.zip');
+        return fs.writeFile(testFileName, 'test')
+          .then(() =>
+            build({
```
same coding style issue mentioned in one of the previous comments.

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,44 @@ describe('build', () => {
+          .then((buildResult) => {
+            assert.match(buildResult.extensionPath,
+                         /minimal_extension-1\.0\.zip$/);
+            return buildResult.extensionPath;
```
This return is probably not really needed.

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +369,41 @@ describe('build', () => {
 
+  describe('defaultPackageCreator', () => {
+
+    it('should reject on Unexpected errors', () => withTempDir(
```
we should keep the coding style of this test consistent with the other tests in the same file:
```js
it('should reject on Unexpected errors', () => {
  return withTempDir(
     (tmpDir) => {
       ...
     });
});
```

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +369,41 @@ describe('build', () => {
+      (tmpDir) => {
+        const fakeEventToPromise = sinon.spy(async (stream) => {
+          await defaultEventToPromise(stream, 'close');
+          const files = await fs.readdir(tmpDir.path());
```
we should add an inline comment here to explain why we are removing the content of the tmpDir here.

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +369,41 @@ describe('build', () => {
+            assert.match(error.message, /Unexpected error/);
+          });
+      }
+  ));
```
This looks like it should be indented with 2 more spaces.

harikishen replied:
the linter tells the indentation is fine, tried adding two but caused error

harikishen commented:
@rpl  i've made the said changes

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +369,42 @@ describe('build', () => {
+  describe('defaultPackageCreator', () => {
+    it('should reject on Unexpected errors', () => {
+      return withTempDir(
+      (tmpDir) => {
```
This should be indented as the other tests in this file.

harikishen replied:
hope it's fine now

rpl commented on src/cmd/build.js:
```diff
@@ -124,11 +139,29 @@ async function defaultPackageCreator({
   const extensionPath = path.join(artifactsDir, packageName);
-  const stream = createWriteStream(extensionPath);
+
+  //Added 'wx' flags to avoid overwriting of existing package
```
Nit, The comments should have a space between `//` and the comment, and ends with a '.'

rpl commented on src/cmd/build.js:
```diff
@@ -124,11 +139,29 @@ async function defaultPackageCreator({
+  try {
+    await eventToPromise(stream, 'close');
+  } catch (error) {
+    if (isErrorWithCode('EEXIST', error)) {
```
@harikishen By re-looking at this block I think that we can reduce the nesting of the `if` blocks by reversing the check and make it nicer:

```js
} catch (error) {
  if (!isErrorWithCode('EEXIST', error)) {
    throw error;
  }

  if (!overwriteDest) {
    ...
  }

  log.info(`Destination exists, overwriting: ${extensionPath}`);
  ...
} 
```

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,43 @@ describe('build', () => {
+  it('raises an UsageError if zip file exists', () => {
+    return withTempDir(
+      (tmpDir) => {
+        const testFileName =
```
This needs to be reformatted as the similar one that is at line 346

rpl commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +368,42 @@ describe('build', () => {
+        (tmpDir) => {
+          const fakeEventToPromise = sinon.spy(async (stream) => {
+            await defaultEventToPromise(stream, 'close');
+            //Remove contents of tmpDir before removal of directory
```
Same coding style issue of one of the previous comments (Add a space between `//` and the comment and ends it with a `.`).

harikishen replied:
@rpl done

harikishen commented:
@rpl done i guess

kumar303 commented on src/program.js:
```diff
@@ -287,6 +287,10 @@ Example: $0 --help run.
           describe: 'Watch for file changes and re-build as needed',
           type: 'boolean',
         },
+        'overwrite-dest': {
```
Can you add `alias: 'o'` to this? This will be a commonly used option so I think it needs an alias.

kumar303 commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -329,4 +367,42 @@ describe('build', () => {
+          };
+
+          return defaultPackageCreator(params, options)
+            .then(() => assert.fail('The expected error has not been raised'))
```
You can import [this helper](https://github.com/mozilla/web-ext/blob/master/tests/unit/helpers.js#L116) and use it like `.then(makeSureItFails()).catch(...)`

kumar303 commented on tests/unit/test-cmd/test.build.js:
```diff
@@ -320,6 +322,42 @@ describe('build', () => {
+              artifactsDir: tmpDir.path(),
+            }))
+          .catch ((error) => {
+            assert.instanceOf(error, UsageError);
```
In addition to asserting the error type, this should also match something specific about the exception message because there are a lot of possible `UsageError`s that can be raised. It doesn't have to match the full message. For example:
````js
assert.match(error.message, /Extension exists at the destination path/);
````

harikishen replied:
@kumar303 I've made the changes you requested
